src/main/java/org/skriptlang/skript/bukkit/tags/TagRegistry.java

```diff
@@ -0,0 +1,206 @@
+package org.skriptlang.skript.bukkit.tags;
+
+import ch.njol.util.coll.iterator.CheckedIterator;
+import com.destroystokyo.paper.MaterialTags;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Iterators;
+import io.papermc.paper.tag.EntityTags;
+import org.bukkit.Keyed;
+import org.bukkit.Material;
+import org.bukkit.NamespacedKey;
+import org.bukkit.Tag;
+import org.bukkit.entity.EntityType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.skriptlang.skript.bukkit.tags.sources.BukkitTagSource;
+import org.skriptlang.skript.bukkit.tags.sources.PaperTagSource;
+import org.skriptlang.skript.bukkit.tags.sources.SkriptTagSource;
+import org.skriptlang.skript.bukkit.tags.sources.TagOrigin;
+import org.skriptlang.skript.bukkit.tags.sources.TagSource;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+/**
+ * A class in charge of storing and handling all the tags Skript can access.
+ */
+public class TagRegistry {
+
+	private final TagSourceMap tagSourceMap = new TagSourceMap();
+
+	/**
+	 * Each new instance will create a new set of tag sources, in an effort to be reload safe.
+	 */
+	TagRegistry() {
+		tagSourceMap.put(TagType.ITEMS, new BukkitTagSource<>("items", TagType.ITEMS));
+		tagSourceMap.put(TagType.BLOCKS, new BukkitTagSource<>("blocks", TagType.BLOCKS));
+		tagSourceMap.put(TagType.ENTITIES, new BukkitTagSource<>("entity_types", TagType.ENTITIES));
+
+		if (TagModule.PAPER_TAGS_EXIST) {
+			try {
+				List<Tag<Material>> materialTags = new ArrayList<>();
+				for (Field field : MaterialTags.class.getDeclaredFields()) {
+					if (field.canAccess(null))
+						//noinspection unchecked
+						materialTags.add((Tag<Material>) field.get(null));
+				}
+				PaperTagSource<Material> paperMaterialTags = new PaperTagSource<>(materialTags, TagType.BLOCKS, TagType.ITEMS);
```

Comment:
Should this filter based on whether the materials represented by the tag are `Material#isItem`, `Material#isBlock`, or both?

Reply:
I don't think that's always safe to do
A Material can be isBlock, but also be a part of an ITEMS tag.
The tag types are conceptual, rather than directly related to what's in them. Maybe we can filter out the non-blocks from block tags, though, let me know what you think.

Reply:
Are there any materials part of an ITEMS tag such that `Material#isItem` is false? Does paper refer to them as items or blocks anywhere?

I just am slightly concerned as I would expect any entries of the `item` tag type to be representable as an ItemStack